pkg/wal/wal.go

```diff
@@ -0,0 +1,345 @@
+	go w.issueParallelReads(ctx, walChannels)
+
+	// Read tokens and send to routine to entry in parallel while fetching new tokens.
+	for count = 0; count < max; {
+		var tokens []string
+
+		tokens, next, err := w.ListTokens(ctx, fromToken, max-count)
+
+		if err != nil {
+			return nil, "", fmt.Errorf("wal failed to ListEntries: " + err.Error())
+		}
+
+		count += len(tokens)
+
+		if len(tokens) > 0 {
+			// account for case list is not complete but the response does not include any token.
+			walChannels.tokens <- tokens
+		}
+
+		fromToken = next
+
+		// Read max or all
+		if (count == max) || (fromToken == "") {
+			break
+		}
+	}
+
+	close(walChannels.tokens) // No more tokens to report.
+
+	walChannels.count <- count // Send total count for reading responses.
+
+	for {
+		select {
+		case entries := <- walChannels.entries:
+			if len(entries) != count {
+				panic(fmt.Sprintf("received count different than the list of tokens entries: %d count: %d", len(entries), count))
+			}
+			return entries, fromToken, nil
+		case err := <- walChannels.oops:
+			return nil, "", err
+		}
+	}
+}
+
+func (w *WAL) issueParallelReads(ctx context.Context, channels *walChannels)  {
+	count := 0
+	for {
+		select {
+		case tokens, ok := <- channels.tokens:
+			if !ok {
+				// Closed and empty: https://golang.org/ref/spec#Receive_operator
+				w.l.Info("Completed issuing parallel reads", zap.Int("count", count))
+				return
+			}
+			for _, t := range tokens {
+				w.getConnection() // concurrency control
+				count++
+				go w.read(ctx, t, channels)
+			}
+		}
+	}
+}
+
+func (w *WAL) read(ctx context.Context, token string, channels *walChannels) {
+	defer w.releaseConnection() // concurrency control
+	r, err := w.walStore.Get(ctx, token)
+	defer r.Close()
+	if err != nil {
+		channels.oops <- err
+		return
+	}
+	var b []byte
+	_, err = io.ReadFull(r, b)
+	entry, err := Unmarshal(b)
+	if err != nil {
+		channels.oops <- fmt.Errorf("token: %s, err: %s", token, err)
+		return
+	}
+	channels.entry <- entry
+	return
+}
+
+func (w *WAL) collectParallelResponses(ctx context.Context, channels *walChannels) {
+	entriesMap := iradix.New().Txn()
+	count := 0
+	total := 0
+	var err error
+
+	finalize := func() {
+		if err != nil {
+			// Report an error occurred.
+			channels.oops <- err
+			return
+		}
+		var entries []Entry
+		iterator := entriesMap.Root().Iterator()
+		for {
+			_, e, ok := iterator.Next()
+			if !ok {
+				// send the final entries
+				channels.entries <- entries
+				return
+			}
+			entry := e.(*Entry)
+			entries = append(entries, *entry)
+		}
+	}
+
+	for {
+		// Wait for all reads to return.
+		select {
+		case e := <- channels.entry:
+			_, updated := entriesMap.Insert([]byte(e.Token), e)
+			if updated {
+				panic(fmt.Sprintf("Received more than one response for token: %s string: %s", e.Token, e.Payload))
+			}
+			count++
+			if count == total {
+				finalize()
+				return
+			}
+		case total = <- channels.count:
```

Comment:
haven't fully verified, yet it appears adding the `finalize` conditional to this case as well could ensure correctness?

impression is that it hasn't been necessary to add because the latency thru `walStore` is more than the time from

```
walChannels.tokens <- tokens
```

to

```
walChannels.count <- count
```

in `ListEntries` ... and while this is probably true in all use-cases, i'm in favor of removing such a tacit assumption (if that's even what's going on in the first place 0_o).

Reply:
Yes that is correct, fixed.